# Practical 1: Load and Explore the Unstructured Access Log Data

**Objective:** Learn how to import and explore raw server access logs (Apache/Nginx format).

- Load a sample `access.log` file using Python
- Understand the typical log structure (IP, timestamp, request, status, size, user-agent, etc.)
- Extract 5-10 random lines to manually inspect the pattern
- Identify useful fields for analysis

In [1]:
import random

with open('logs/access.log', 'r') as file:
    log_lines = file.readlines()

print("Total number of log lines:", len(log_lines))

Total number of log lines: 47685


## First log entry

In [2]:
print(log_lines[0])

144.187.77.221 - - [01/Aug/2026:00:00:25 +0530] "GET /api/products HTTP/1.1" 404 418 "https://facebook.com" "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"



## Useful fields in Access Logs

A standard Apache/Nginx **Combined Log Format** entry contains:

| Field | Example | Useful for |
|---|---|---|
| IP Address | `192.168.1.10` | tracking source, rate-limiting, geo |
| Timestamp | `[12/Aug/2026:10:15:32 +0530]` | time-series analysis, brute-force detection |
| HTTP Method | `GET` / `POST` | request type analysis |
| Resource | `/index.html` | attack pattern detection (SQLi, traversal) |
| Protocol | `HTTP/1.1` | sanity check |
| Status Code | `200`, `404`, `500` | anomaly detection |
| Bytes Sent | `4521` | payload size anomalies |
| Referrer | `https://google.com` | traffic source |
| User-Agent | `Mozilla/5.0 ...` | bot vs. real browser detection |

## Random sample of 5 log entries

In [3]:
random_logs = random.sample(log_lines, 5)

print("Random Log Entries:")
for entry in random_logs:
    print(entry.strip())

Random Log Entries:
181.87.155.170 - - [02/Aug/2026:15:50:40 +0530] "GET / HTTP/1.1" 200 8109 "-" "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"
60.55.183.74 - - [10/Aug/2026:15:50:46 +0530] "GET /dashboard HTTP/1.1" 404 6920 "https://bing.com" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"
107.114.189.22 - - [05/Aug/2026:09:28:19 +0530] "GET /checkout HTTP/1.1" 200 5876 "https://bing.com" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"
129.129.158.110 - - [14/Aug/2026:03:46:12 +0530] "GET /config.php HTTP/1.1" 404 391 "-" "python-requests/2.31.0"
143.189.132.85 - - [13/Aug/2026:03:59:54 +0530] "GET /login HTTP/1.1" 404 5263 "https://example.com" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"


## Observations

- Each line follows the Combined Log Format: `IP - - [timestamp] "METHOD path PROTOCOL" status bytes "referrer" "user-agent"`
- Fields are space-separated, but the request, timestamp, referrer, and user-agent are each wrapped in `[]` or `""` since they may contain spaces themselves — this is why regex (not `.split()`) will be needed to parse them reliably in Practical 2.
- The most analytically useful fields for anomaly/attack detection later are: **Resource** (for SQLi/traversal patterns), **Status Code** (for error/anomaly spikes), and **IP + Timestamp** (for brute-force / rate-based detection).